# HG changeset patch
# User John W. Eaton <jwe@octave.org>
# Date 1365696164 14400
#      Thu Apr 11 12:02:44 2013 -0400
# Node ID d3b01ae21259d3f57fd233b727a065cd18a6045c
# Parent  b98ad20e48802d5699dbcb5394d4797552441013
temporary fix for terminal window resizing

* input.cc (octave_base_reader::octave_gets): Call
command_editor::reset_screen_size.

* cmd-edit.h, cmd-edit.cc (command_editor::reset_screen_size,
command_editor::do_reset_screen_size,
gnu_readline::do_reset_screen_size): New functions.
* oct-rl-edit.h, oct-rl-edit.c (octave_rl_reset_screen_size):
New function.

diff --git a/libinterp/interpfcn/input.cc b/libinterp/interpfcn/input.cc
--- a/libinterp/interpfcn/input.cc
+++ b/libinterp/interpfcn/input.cc
@@ -215,16 +215,20 @@ octave_base_reader::octave_gets (bool& e
   if (interactive || forced_interactive)
     {
       if (! Vdebugging)
         octave_link::exit_debugger_event ();
 
       octave_link::pre_input_event ();
 
       octave_link::set_workspace (symbol_table::workspace_info ());
+
+      // FIXME -- this call should happen any time the terminal window
+      // size changes, not just prior to prompting for input.
+      command_editor::reset_screen_size ();
     }
 
   bool history_skip_auto_repeated_debugging_command = false;
 
   std::string ps = (pflag > 0) ? VPS1 : VPS2;
 
   std::string prompt = command_editor::decode_prompt_string (ps);
 
diff --git a/liboctave/util/cmd-edit.cc b/liboctave/util/cmd-edit.cc
--- a/liboctave/util/cmd-edit.cc
+++ b/liboctave/util/cmd-edit.cc
@@ -90,16 +90,18 @@ public:
   int do_terminal_rows (void);
 
   int do_terminal_cols (void);
 
   void do_clear_screen (void);
 
   void do_resize_terminal (void);
 
+  void do_reset_screen_size (void);
+
   std::string newline_chars (void);
 
   void do_restore_terminal_state (void);
 
   void do_blink_matching_paren (bool flag);
 
   void do_set_basic_word_break_characters (const std::string& s);
 
@@ -326,16 +328,22 @@ gnu_readline::do_clear_screen (void)
 }
 
 void
 gnu_readline::do_resize_terminal (void)
 {
   ::octave_rl_resize_terminal ();
 }
 
+void
+gnu_readline::do_reset_screen_size (void)
+{
+  ::octave_rl_reset_screen_size ();
+}
+
 std::string
 gnu_readline::newline_chars (void)
 {
   return "\r\n";
 }
 
 void
 gnu_readline::do_restore_terminal_state (void)
@@ -989,16 +997,23 @@ command_editor::clear_screen (void)
 
 void
 command_editor::resize_terminal (void)
 {
   if (instance_ok ())
     instance->do_resize_terminal ();
 }
 
+void
+command_editor::reset_screen_size (void)
+{
+  if (instance_ok ())
+    instance->do_reset_screen_size ();
+}
+
 std::string
 command_editor::decode_prompt_string (const std::string& s)
 {
   return (instance_ok ())
     ? instance->do_decode_prompt_string (s) : std::string ();
 }
 
 int
diff --git a/liboctave/util/cmd-edit.h b/liboctave/util/cmd-edit.h
--- a/liboctave/util/cmd-edit.h
+++ b/liboctave/util/cmd-edit.h
@@ -74,16 +74,18 @@ public:
   static int terminal_rows (void);
 
   static int terminal_cols (void);
 
   static void clear_screen (void);
 
   static void resize_terminal (void);
 
+  static void reset_screen_size (void);
+
   static std::string decode_prompt_string (const std::string& s);
 
   static void restore_terminal_state (void);
 
   static void blink_matching_paren (bool flag);
 
   static void set_basic_word_break_characters (const std::string& s);
 
@@ -216,16 +218,18 @@ protected:
   virtual int do_terminal_rows (void) { return 24; }
 
   virtual int do_terminal_cols (void) { return 80; }
 
   virtual void do_clear_screen (void) { }
 
   virtual void do_resize_terminal (void) { }
 
+  virtual void do_reset_screen_size (void) { }
+
   virtual std::string do_decode_prompt_string (const std::string&);
 
   virtual std::string newline_chars (void) { return "\n"; }
 
   virtual void do_restore_terminal_state (void) { }
 
   virtual void do_blink_matching_paren (bool) { }
 
diff --git a/liboctave/util/oct-rl-edit.c b/liboctave/util/oct-rl-edit.c
--- a/liboctave/util/oct-rl-edit.c
+++ b/liboctave/util/oct-rl-edit.c
@@ -106,16 +106,22 @@ octave_rl_clear_screen (void)
 
 void
 octave_rl_resize_terminal (void)
 {
   rl_resize_terminal ();
 }
 
 void
+octave_rl_reset_screen_size (void)
+{
+  rl_reset_screen_size ();
+}
+
+void
 octave_rl_restore_terminal_state ()
 {
   if (rl_deprep_term_function)
     rl_deprep_term_function ();
 }
 
 void
 octave_rl_insert_text (const char *s)
diff --git a/liboctave/util/oct-rl-edit.h b/liboctave/util/oct-rl-edit.h
--- a/liboctave/util/oct-rl-edit.h
+++ b/liboctave/util/oct-rl-edit.h
@@ -55,16 +55,18 @@ extern int octave_rl_screen_width (void)
 extern void octave_rl_enable_paren_matching (int);
 
 extern void octave_rl_init (void);
 
 extern void octave_rl_clear_screen (void);
 
 extern void octave_rl_resize_terminal (void);
 
+extern void octave_rl_reset_screen_size (void);
+
 extern void octave_rl_restore_terminal_state (void);
 
 extern void octave_rl_insert_text (const char *);
 
 extern int octave_rl_newline (int, int);
 
 extern const char *octave_rl_line_buffer (void);
 
